pkg/middleware/middleware.go

```diff
@@ -0,0 +1,48 @@
+package middleware
+
+import (
+	"fmt"
+	"github.com/gazebo-web/auth/pkg/authentication"
+	"github.com/golang-jwt/jwt/v5/request"
+	"net/http"
+)
+
+// Extractor extracts a string value from an HTTP request. It's usually used to extract a header from an HTTP request,
+// but can also be used for extracting a user and password from the body.
+//
+// There are a few implementations already provided by the request package, for example:
+// Bearer tokens: request.BearerExtractor
+type Extractor = request.Extractor
+
+// Middleware is used to modify or augment the behavior of an HTTP request handler.
+type Middleware func(http.Handler) http.Handler
```

Comment:
Why does a middleware return a function that returns an HTTP handler as opposed to being an HTTP handler itself?

There might be a use case or intended usage that I'm missing.

Reply:
A middleware returns any type that fulfills the Handler interface, it happens that I'm using `http.HandlerFunc` as the implementation for that interface in this code.

The idea behind a middleware is to generate a new handler that changes the original behavior of the HTTP handler, so that's the reason why the signature of this function receives a handler, and returns the modified handler.

Reply:
> The idea behind a middleware is to generate a new handler that changes the original behavior of the HTTP handler, so that's the reason why the signature of this function receives a handler, and returns the modified handler.

If I'm understanding correctly, this is how you would end up using this middleware.

```go
// Assume this is configured somehow
auth := NewAuth(...)

bearerMW := BearerToken(auth)

// Assume handlers were previously defined
handler1 = bearerMW(handler1)
handler2 = bearerMW(handler2)
handler3 = bearerMW(handler3)
```

If we remove the `Middleware` type altogether and just have middleware functions instead, then we remove the step of having to create the middleware and apply it. This does require passing the necessary parameters to each middleware _each time_, but I don't expect this to be an issue.

I'm proposing that we remove the`Middleware` type and have the functions act as decorators/wrappers.

`BearerToken(authentication.Authentication, http.Handler) http.Handler`

Usage would then look something like this

```go
// Assume this is configured somehow
auth := NewAuth(...)

// Assume handlers were previously defined
handler1 = BearerToken(auth, handler1)
handler2 = BearerToken(auth, handler1)
handler3 = BearerToken(auth, handler1)
```

I don't feel too strongly about this, but I'm trying to understand if we can reduce the number of types and setup required to use our middlewares.

Reply:
I based the original proposal from these two frameworks, and following a higher-order func approach that allows you to configure a middleware:

https://github.com/gorilla/mux/blob/master/middleware.go#L11
https://github.com/go-chi/chi/blob/master/middleware/strip.go#L13

I'd like to stress out that we're not covering production-grade use cases with this discussion, as most of the frameworks let you set up a middleware using a `Use` method that accepts a middleware function (each framework has its own way of configuring it) and you're not supposed to pass custom data in the middleware signature

- https://gin-gonic.com/zh-tw/docs/examples/using-middleware/
- https://go-chi.io/#/pages/middleware
- https://docs.gofiber.io/v1.x/api/middleware/

Additionally, I think that it's easier to read when using specific middleware funcs instead of decorating handlers:

```go
handler := NewDoSomethingHandler(...)
auth := NewAuth(...)
logger := NewLogger(...)
tracing := NewTracing(...)


handler = BearerToken(auth, Logger(logger, Tracing(tracing, handler))
```

But with the original proposal:

```go
// Initialize handlers
handler := NewDoSomethingHandler(...)

// Initialize services
authSvc := NewAuth(...)
loggerSvc := NewLogger(...)
tracingSvc := NewTracing(...)

// Middlewares

auth := BearerToken(authSvc)
logger := Logger(loggerSvc)
tracing := Tracing(tracingSvc)

// Middlewares
handler = auth(logger(tracing(handler)))
```

Reply:
Fair enough. Appreciate you taking the time to discuss this.